### PR TITLE
Spec attribution rate-limiting

### LINKED ISSUE #246: Spec attribution rate-limiting

### PR DESCRIPTION
**Pull request by apasel422**
Fixes #246

***
<a href="https://pr-preview.s3.amazonaws.com/apasel422/conversion-measurement-api/pull/253.html" title="Last updated on Nov 9, 2021, 5:54 PM UTC (b20f603)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/conversion-measurement-api/253/ee4f982...apasel422:b20f603.html" title="Last updated on Nov 9, 2021, 5:54 PM UTC (b20f603)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -680,6 +698,15 @@ To <dfn>trigger attribution</dfn> given an [=attribution trigger=] |trigger| run
     [=attribution report/attribution destination=] equals |attributionDestination|.
 1. If |numMatchingReports| is greater than or equal to the user agent's
     [=max reports per attribution destination=], return.
+1. Set |rateLimitCutoff| to an [=implementation-defined=] time before |trigger|'s [=attribution trigger/trigger time=].
```
Comment (johnivdel) [inline index.bs:701]: Could we factor this into a separate algorithm?

File: index.bs [hunk 2]
```diff
@@ -706,6 +733,17 @@ To <dfn>trigger attribution</dfn> given an [=attribution trigger=] |trigger| run
...
+
+    : [=attribution rate-limit record/source site=]
+    :: |sourceSite|
+    : [=attribution rate-limit record/attribution destination=]
+    :: |attributionDestination|
+    : [=attribution rate-limit record/source type=]
+    :: |sourceToAttribute|'s [=attribution source/source type=]
+    : [=attribution rate-limit record/trigger time=]
+    :: |trigger|'s [=attribution trigger/trigger time=]
+1. Add |rateLimitRecord| to the [=attribution rate-limit cache=].
```
Comment (johnivdel) [inline index.bs:746]: Do we want to note removal of these entries somewhere? E.g. the user agent may remove all record whose trigger time is less than now - cutoff

File: index.bs [hunk 3]
```diff
@@ -680,6 +698,15 @@ To <dfn>trigger attribution</dfn> given an [=attribution trigger=] |trigger| run
...
 1. If |numMatchingReports| is greater than or equal to the user agent's
     [=max reports per attribution destination=], return.
+1. Set |rateLimitCutoff| to an [=implementation-defined=] time before |trigger|'s [=attribution trigger/trigger time=].
+1. Let |sourceSite| be the result of running [=parse a site=] with |sourceToAttribute|'s [=attribution source/source origin=].
+1. If |sourceSite| is null, return.
+1. Let |matchingRateLimitRecords| be all entries in the [=attribution rate-limit cache=] where all of the following are true:
+     * entry's [=attribution rate-limit record/source site=] and |sourceSite| are equal
+     * entry's [=attribution rate-limit record/attribution destination=] and |attributionDestination| are equal
+     * entry's [=attribution rate-limit record/source type=] and |sourceToAttribute|'s [=attribution source/source type=] are equal
+     * entry's [=attribution rate-limit record/trigger time=] is greater than |rateLimitCutoff|
```
Comment (johnivdel) [inline index.bs:708]: rateLimitTimePeriod / rateLimitWIndowLength or something might make this variable a bit more clear

**Review by johnivdel — COMMENTED** [review]
Just a few comments. Thanks for specing the full algorithm here.

**Comment by apasel422** [comment]
PTAL

File: index.bs [hunk 4]
```diff
@@ -670,6 +688,28 @@ Given a [=request=] |request|:
 1. [=Queue a task=] to [=trigger attribution=] with |trigger|.
 1. Return <strong>blocked</strong>.
 
+<h3 dfn id="should-rate-limit-attribution">Should attribution be rate-limited</h3>
```
Comment (johnivdel) [inline index.bs:691]:
```suggestion
<h3 dfn id="should-rate-limit-attribution">Should attribution be blocked by rate limit</h3>
```

**Review by johnivdel — APPROVED** [review]

**Pull request merged**